### DISCUSSION TITLE
Idea for: SDXL TypeError: '<' not supported between instances of 'NoneType' and 'int' (for prompt)... Discussion.

### DISCUSSION THREAD

**Discussion by DirtyHamster** [vladmandic/sdnext#1706] [Ideas]
Don't consider this an error report. As it's really easy to see what's causing it. I would have chimed this in on discord but I don't have that set up right now (Haven't gotten around to reinstalling it yet, had uninstalled it from never using it). 

So I had been getting an error when I first started Playing around with SDXL's implementation and it turned out that it couldn't deal with not having a prompt in the Refiner fields. It's pretty clear on that when it happens.

 H:\Users\adamf\AI_Progs\VladDiffusion\automatic\venv\lib\site-packages\diffusers\pipelines\stable_diffusion_xl\pipel │
│ ine_stable_diffusion_xl_img2img.py:740 in __call__                                                                   │
│                                                                                                                      │
│   739 │   │   # 1. Check inputs. Raise error if not correct                                                          │
│ ❱ 740 │   │   self.check_inputs(prompt, strength, callback_steps, negative_prompt, prompt_embe                       │
│   741

I was thinking a really nice fix for this might be when that field is 'NoneType' that it should probably just by default pull the primary prompt to solve for the error. I could also see it where you have an optional default that you could set too and I'm not sure which one would be better or if there is some even better idea out there for it. So I wanted to open up a discussion on what I think is actually an interesting basic error. I could see a few of these simple fixes being implemented with some simple arg switches in the UI.

**Comment by kaneda2004** [comment]
I am having this issue as well, it appears in the following code block - I believe the strength variable is None, therefore it throws a value error. It seems that the denoising strength variable is not being passed to this function. @vladmandic 

```
    def check_inputs(
        self, prompt, strength, callback_steps, negative_prompt=None, prompt_embeds=None, negative_prompt_embeds=None
    ):
        if strength < 0 or strength > 1:
            raise ValueError(f"The value of strength should in [0.0, 1.0] but is {strength}")

        if (callback_steps is None) or (
            callback_steps is not None and (not isinstance(callback_steps, int) or callback_steps <= 0)
```

**Reply by DirtyHamster** [reply]
You can just copy down your prompt from the main prompt to fix it. 

That's This one:

self, prompt, strength, callback_steps, negative_prompt=None, prompt_embeds=None, negative_prompt_embeds=None

The second one is Denoise start and end just leave them at 0 and 1.

 if strength < 0 or strength > 1:
            raise ValueError(f"The value of strength should in [0.0, 1.0] but is {strength}")
            
Third one I haven't figured out yet but have gotten a few times it's a less then sign but I'm not sure if that is looking at steps to steps between int sections or something else.  

 if (callback_steps is None) or (
            callback_steps is not None and (not isinstance(callback_steps, int) or callback_steps <= 0)  
            
If we could focus on the first one though to stay on topic that one is very repeatable. As I was saying there are options to solve it in a number of ways. We should probably list them so we can pick threw what would be more useful. I.e. adding a prompt clears it but that can be done in a number of ways.

As Vlad was saying he wants to do the trouble shooting on discourse but I'm not on there I'll try to deal with that later in the week. I don't want this topic to detract from that. If someone can confirm the first one as a fix that would be good though. We can have a meaningful discussion based on what might also work for it,

**Comment by kaneda2004** [comment]
Sorry, I should have articulated further:
I do have both prompts filled out. 
I do have denoising set correctly, with all of these values set, the value for the denoising strength is not being passed to that specific function. - strangely it's intermittent - depending on the current state of the server. I'll have to dig deeper.

**Comment by vladmandic** [comment]
> As it's really easy to see what's causing it

actually, its not clear to me at all.
refiner prompt already has that check: `prompt=[p.refiner_prompt] if len(p.refiner_prompt) > 0 else prompts`
and `strength` gets value from ui **if** second pass is enabled or we're in img2img.

so what is passed as `None` and under which exact workflow?

**Reply by kaneda2004** [reply]
I believe the denoising strength `strength` param is occasionally passed as `None` to the `check_inputs` function in pipeline_stable_diffusion_xl_img2img.py

It has intermittently thrown an error for me, that only re-launching the server resolves - I'm unable at this point to determine what causes it to occur - sometimes it has just happened right after a successful generation. 

Is there any piece of code that's resetting the state of the `strength` variable?

Also, I believe @DirtyHamster is indicating that if the second pass positive prompt is left blank, it similarly will throw an error for the None value being passed. I also agree with @DirtyHamster that the primary prompt could be auto-populated into the second pass prompt (perhaps by using a checkbox so the user has control still).

Could bypass the checks in pipeline_stable_diffusion_xl_img2img.py if they're already checked within your code?

Thanks for taking a look

Console exception log below:

`17:41:29-862743 ERROR    gradio call: TypeError
╭───────────────────────────────────────── Traceback (most recent call last) ──────────────────────────────────────────╮
│ F:\StableD\automatic2\automatic\modules\call_queue.py:34 in f                                                        │
│                                                                                                                      │
│    33 │   │   │   try:                                                                                               │
│ ❱  34 │   │   │   │   res = func(*args, **kwargs)                                                                    │
│    35 │   │   │   │   progress.record_results(id_task, res)                                                          │
│                                                                                                                      │
│ F:\StableD\automatic2\automatic\modules\txt2img.py:63 in txt2img                                                     │
│                                                                                                                      │
│   62 │   if processed is None:                                                                                       │
│ ❱ 63 │   │   processed = processing.process_images(p)                                                                │
│   64 │   p.close()                                                                                                   │
│                                                                                                                      │
│                                               ... 3 frames hidden ...                                                │
│                                                                                                                      │
│ F:\StableD\automatic2\automatic\venv\lib\site-packages\diffusers\pipelines\stable_diffusion_xl\pipeline_stable_diffu │
│ sion_xl_img2img.py:740 in __call__                                                                                   │
│                                                                                                                      │
│   739 │   │   # 1. Check inputs. Raise error if not correct                                                          │
│ ❱ 740 │   │   self.check_inputs(prompt, strength, callback_steps, negative_prompt, prompt_embe                       │
│   741                                                                                                                │
│                                                                                                                      │
│ F:\StableD\automatic2\automatic\venv\lib\site-packages\diffusers\pipelines\stable_diffusion_xl\pipeline_stable_diffu │
│ sion_xl_img2img.py:460 in check_inputs                                                                               │
│                                                                                                                      │
│   459 │   ):                                                                                                         │
│ ❱ 460 │   │   if strength < 0 or strength > 1:                                                                       │
│   461 │   │   │   raise ValueError(f"The value of strength should in [0.0, 1.0] but is {streng
TypeError: '<' not supported between instances of 'NoneType' and 'int'
17:51:28-292298 INFO     Exiting`

**Comment by DirtyHamster** [comment]
Running the exact set up and inputting a prompt removed the first error.  Standard SD 1.5 using diffusers with the SDXL refiner open.  Select diffuser pipeline when loading from safetensors, using the new diffusers.  No vae that was causing additional errors. I can reproduce that though as it happens pretty much every time I open a vae.

Img2img is what the refiner is using on the later stage second pass. In the independent panel for the refiner that has to be filled out that prompt is causing the issue. As a check If I remove the refiner the error also goes away...
  
See image:
![image](https://github.com/vladmandic/automatic/assets/71010673/77b7f6b6-48f7-4a89-bf2a-f4eab7140a14)

Prompt in question is below the one causing the none type prompt error. it might have been intended to grab the default prompt looking at the else statement but it might just be pointed wrong. Is the main prompt really just called prompts, I was thinking that would be more defined pathwise.

The denoise start and stop in the image causes the second error any time I vary those I get the same error pretty much anytime I change the value. In this case the error states what it expects: The value of strength should in [0.0, 1.0]

Ok for the error codes as a recurring event I just removed the prompts from that section in the pic the full feed out is the following:

AttributeError: 'NoneType' object has no attribute 'repeat'
Progress 37.29s/it ██████████████████████████████████████ 100% 20/20 12:25 00:00
23:17:24-988986 WARNING  GPU high memory utilization: 100% {'ram': {'used': 4.02, 'total': 31.92}, 'gpu': {'used':
                         11.99, 'total': 11.99}, 'retries': 0, 'oom': 0}     /// ----- it never oom's lol ..... ///
23:17:26-141127 ERROR    Exception: 'NoneType' object has no attribute 'repeat'
23:17:26-142126 ERROR    Arguments: args=('task(uxs0ppk382spmk0)',
The actual main positive prompt sits here... 
End of main negative, then this comes up:
', [], 20, 0, 0, False, False, 1, 1, 5,
                         1, 0.5, 1, -1.0, -1.0, 0, 0, 0, 1024, 1024, True, 0.5, 2, 'Latent', 20, 0, 0, 0, 1, '', '', [],
                         0, False, {'ad_model': 'face_yolov8n.pt', 'ad_prompt': '', 'ad_negative_prompt': '',
                         'ad_confidence': 0.3, 'ad_mask_min_ratio': 0, 'ad_mask_max_ratio': 1, 'ad_x_offset': 0,
                         'ad_y_offset': 0, 'ad_dilate_erode': 4, 'ad_mask_merge_invert': 'None', 'ad_mask_blur': 4,
                         'ad_denoising_strength': 0.4, 'ad_inpaint_only_masked': True, 'ad_inpaint_only_masked_padding':
                         32, 'ad_use_inpaint_width_height': False, 'ad_inpaint_width': 512, 'ad_inpaint_height': 512,
                         'ad_use_steps': False, 'ad_steps': 28, 'ad_use_cfg_scale': False, 'ad_cfg_scale': 7,
                         'ad_use_noise_multiplier': False, 'ad_noise_multiplier': 1, 'ad_restore_face': False,
                         'ad_controlnet_model': 'None', 'ad_controlnet_module': 'inpaint_global_harmonious',
                         'ad_controlnet_weight': 1, 'ad_controlnet_guidance_start': 0, 'ad_controlnet_guidance_end': 1,
                         'is_api': ()}, {'ad_model': 'None', 'ad_prompt': '', 'ad_negative_prompt': '', 'ad_confidence':
                         0.3, 'ad_mask_min_ratio': 0, 'ad_mask_max_ratio': 1, 'ad_x_offset': 0, 'ad_y_offset': 0,
                         'ad_dilate_erode': 4, 'ad_mask_merge_invert': 'None', 'ad_mask_blur': 4,
                         'ad_denoising_strength': 0.4, 'ad_inpaint_only_masked': True, 'ad_inpaint_only_masked_padding':
                         32, 'ad_use_inpaint_width_height': False, 'ad_inpaint_width': 512, 'ad_inpaint_height': 512,
                         'ad_use_steps': False, 'ad_steps': 28, 'ad_use_cfg_scale': False, 'ad_cfg_scale': 7,
                         'ad_use_noise_multiplier': False, 'ad_noise_multiplier': 1, 'ad_restore_face': False,
                         'ad_controlnet_model': 'None', 'ad_controlnet_module': 'inpaint_global_harmonious',
                         'ad_controlnet_weight': 1, 'ad_controlnet_guidance_start': 0, 'ad_controlnet_guidance_end': 1,
                         'is_api': ()}, False, False, 'positive', 'comma', 0, False, False, '', 0, '', [], 0, '', [], 0,
                         '', [], True, False, False, False, 0, False) kwargs={}


23:17:26-152127 ERROR    gradio call: AttributeError
╭───────────────────────────────────────── Traceback (most recent call last) ──────────────────────────────────────────╮
│ H:\Users\adamf\AI_Progs\VladDiffusion\automatic\modules\call_queue.py:34 in f                                        │
│                                                                                                                      │
│    33 │   │   │   try:                                                                                               │
│ ❱  34 │   │   │   │   res = func(*args, **kwargs)                                                                    │
│    35 │   │   │   │   progress.record_results(id_task, res)                                                          │
│                                                                                                                      │
│ H:\Users\adamf\AI_Progs\VladDiffusion\automatic\modules\txt2img.py:63 in txt2img                                     │
│                                                                                                                      │
│   62 │   if processed is None:                                                                                       │
│ ❱ 63 │   │   processed = processing.process_images(p)                                                                │
│   64 │   p.close()                                                                                                   │
│                                                                                                                      │
│                                               ... 3 frames hidden ...                                                │
│                                                                                                                      │
│ H:\Users\adamf\AI_Progs\VladDiffusion\automatic\venv\lib\site-packages\diffusers\pipelines\stable_diffusion_xl\pipel │
│ ine_stable_diffusion_xl_img2img.py:766 in __call__                                                                   │
│                                                                                                                      │
│   765 │   │   │   negative_pooled_prompt_embeds,                                                                     │
│ ❱ 766 │   │   ) = self.encode_prompt(                                                                                │
│   767 │   │   │   prompt,                                                                                            │
│                                                                                                                      │
│ H:\Users\adamf\AI_Progs\VladDiffusion\automatic\venv\lib\site-packages\diffusers\pipelines\stable_diffusion_xl\pipel │
│ ine_stable_diffusion_xl_img2img.py:433 in encode_prompt                                                              │
│                                                                                                                      │
│   432 │   │   )                                                                                                      │
│ ❱ 433 │   │   negative_pooled_prompt_embeds = negative_pooled_prompt_embeds.repeat(1, num_imag                       │
│   434 │   │   │   bs_embed * num_images_per_prompt, -1                                                               │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'NoneType' object has no attribute 'repeat'

Give me a sec to put them back in one at a time and you can see how it changes.

**Comment by vladmandic** [comment]
honestly, i've tried reading through this and its mixes what you think is going on vs what is going on and its really hard to follow.

like this
> Also, I believe @DirtyHamster is indicating that if the second pass positive prompt is left blank, it similarly will throw an error for the None value being passed. I also agree with @DirtyHamster that the primary prompt could be auto-populated into the second pass prompt

read my immedate post before that - that is EXACTLY what its doing, using refiner prompt if it exists and populating with normal prompt if it doesnt. i even posted the line of the code.

so please, exact steps to reproduce. everything else is just adding to confusion.

**Reply by kaneda2004** [reply]
1. Clean launch, do not autoload model, backend set to diffusers, pipeline set to Stable Diffusion XL, Refiner model selected
![image](https://github.com/vladmandic/automatic/assets/4466807/d88c41dd-f8b4-4ddf-8796-709a63a01fa7)

2. Enter prompt, hit generate, it works
3. tick Second pass checkbox
4. do not enter refiner prompt
5. hit generate results in this error:
![image](https://github.com/vladmandic/automatic/assets/4466807/31134eae-5727-4497-8498-224c9b265e39)

6. In order to get it working again, the server must be shut down and relaunched

Noted before - it's intermittent - sometimes following the same steps above results in no errors whatsoever.
I will keep investigating.

Edit: Note I just pulled commit e2b33b8, launched server with --debug parameter and it appears to be working as expected. I will see if I can replicate the error on this commit.

Thanks kindly

**Comment by DirtyHamster** [comment]
@vladmandic  I get what your saying and it takes me 10 - 20 minutes to run a test too trying to grab the exact error and work flow. How long would it take to make a copy settings button? i.e something that grabs all the info you want to see for a report? I'm asking because I want to make sure you get the info you need not just what I copy over.. Behavior before the error was normal normal launch, copy prompts, set settings in the prompt window never in the same order but same steps. gen first image and go from there...  This is the only portion that was different I don't normally bother with the hi-res or second pass unless I find a seed I like to use it on. As the refiner was available I figured I'd try it so I enabled second pass. 

I didn't have any errors until I messed with the refiner and it's settings and neglected it's prompts. That's what had me narrowing it down. Not having a prompt without touching any setting in that section had the first error. The error went away after adding a prompt to both pos and neg. I had reproduced that a number of times across the morning. (fresh restarts is probably the only thing I skipped this time. I've been AFK for hours.  

All I had to do to get it to start was leave out the prompt in the second prompt and or change that 2nd error range to above 0 or less then 1. Those were independent errors though. 

@kaneda2004 I pretty much had to close out both the server and browser to clear it each time.

**Comment by ASeale** [comment]
if "upscale by" or "secondary cfg scale" are set to 1 it throws the error. Increasing them to be just above 1 and the error goes away.

**Reply by DirtyHamster** [reply]
The upscaler is disabled as far as the UI says. The secondary CFG scale is in the section that we are mostly talking about.

**Reply by ASeale** [reply]
I know it says disabled but change it above 1.0 anyway. Trust me.

**Reply by ASeale** [reply]
I stumbled across the error as I was changing each setting in my hunt to make the refiner work again, and I even changed the upscaler just to see. It isn't upscaling, but it still looks at that value. Shared code and all.

**Reply by DirtyHamster** [reply]
I got what you meant and I wasn't saying that it doesn't toss an error but it's just not the section we were talking about. Grab the error code next time you can though so we can look at it though.

**Reply by ASeale** [reply]
Certainly. It's the exact same error that this thread references regardless of which one I set to 1.0. Hope this helps. (edit: I noticed after this posted that they are, in fact, different errors. My eyes lied to me.)
<img width="1698" alt="image" src="https://github.com/vladmandic/automatic/assets/21082704/19e44af9-5c4c-4317-a275-f476049c99e4">
<img width="1637" alt="image" src="https://github.com/vladmandic/automatic/assets/21082704/cbf99631-2858-4af9-879b-f7845cf4bc21">
<img width="1608" alt="image" src="https://github.com/vladmandic/automatic/assets/21082704/a42d8c47-daac-4363-9177-0529ff955d99">

**Reply by ASeale** [reply]
I can reproduce your error 100% of the time by setting either to 1.0. Can generate an image 100% of the time keeping those above 1.0 and never entering any refiner prompts. So I don't believe it's tied to the prompts like you were suggesting.

edit: ok, so both are different errors....... dunno why my eyes saw the same error.  However, the error is reproducible using 1.0.

**Comment by DirtyHamster** [comment]
@Aseale
> I can reproduce your error 100%
We have to tell Vlad how to do it step by step. Can you Help with that?

**Comment by ASeale** [comment]
@vladmandic I am able to reproduce the error by simply making the "Upscale By" == 1.0

1. Launch SD.Next using diffuser backend and use SDXL 0.9 (so the HiRes Resize is Disabled)
2. Enter prompt
3. Select "Second Pass"
4. Set "Upscale By" to 1
5. I get the error regardless if I leave the Refiner prompts empty or populate them
6. Hit Generate
7. After the Base steps process and when the Refiner steps should begin you get the error
![image](https://github.com/vladmandic/automatic/assets/21082704/f3fb91e2-5b6a-4727-a67f-431dacf19135)

On another note, you get a similar error if you use the same steps above but change STEP 4 to say "set 'Secondary CFG Scale' to 1":
![image](https://github.com/vladmandic/automatic/assets/21082704/cc793d42-56d9-4474-b187-fced01eae375)

**Reply by ASeale** [reply]
@DirtyHamster and @kaneda2004 I do not have to restart the server to clear this error, I just raise the value to be above 1.0 and then generate again.

**Reply by vladmandic** [reply]
@ASeale exact steps to reproduce help a lot, i've fixed it in dev, will publish in 1-2h.
